### PR TITLE
HS-1476 Prevent creating multiple nodes when receiving trap events

### PR DESCRIPTION
**Pull request by yangli2014**
## Description

- Prevent creating multiple nodes with the trap events from the same tenant, same location, and same IP address
- Prevent triggering an alert before a node is created.

## Jira link(s)
- https://issues.opennms.org/browse/HS-1476

## Checklist
* [ ] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

### CODE REVIEW

File: inventory/main/src/main/java/org/opennms/horizon/inventory/component/NodeMonitoringManager.java [hunk 1]
```diff
@@ -67,14 +71,20 @@ public void receiveTrapEvent(@Payload byte[] data) {
...
-                    .setManagementIp(event.getIpAddress())
-                    .setLabel("trap-" + event.getIpAddress())
-                    .setMonitoredState(MonitoredState.DETECTED)
-                    .build();
-                Node node = nodeService.createNode(createDTO, ScanType.NODE_SCAN, tenantId);
-                passiveDiscoveryService.sendNodeScan(node);
+                Optional<IpInterfaceDTO> ipInterfaceOp = ipInterfaceService.findByIpAddressAndLocationAndTenantId(event.getIpAddress(), event.getLocation(), event.getTenantId());
+                if(ipInterfaceOp.isPresent()) {
+                    IpInterfaceDTO ipInterface = ipInterfaceOp.get();
+                    log.warn("IP address {} already exists in the system and belong to device with ID {}", event.getIpAddress(), ipInterface.getNodeId());
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/component/NodeMonitoringManager.java:77]: what do you think of moving this check to `nodeService.createNode`, that will avoid doing this at multiple places and will have common pattern.

Comment (yangli2014) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/component/NodeMonitoringManager.java:77]: Good idea

File: inventory/main/src/test/java/org/opennms/horizon/inventory/service/NodeServiceTest.java [hunk 2]
```diff
@@ -249,4 +259,18 @@ public void testListNodesByIdsEmpty() {
...
+        Node node = new Node();
+        IpInterface ipInterface = new IpInterface();
+        ipInterface.setNode(node);
+        NodeCreateDTO nodeCreate = NodeCreateDTO.newBuilder()
+            .setLabel("test-node")
+            .setManagementIp("127.0.0.1").build();
+        doReturn(Optional.of(ipInterface)).when(mockIpInterfaceRepository).findByIpAddressAndLocationAndTenantId(any(InetAddress.class), eq(nodeCreate.getLocation()), eq(tenantID));
+        Node result = nodeService.createNode(nodeCreate, ScanType.NODE_SCAN, tenantID);
+        assertThat(result).isEqualTo(node);
+        verify(mockIpInterfaceRepository).findByIpAddressAndLocationAndTenantId(any(InetAddress.class), eq(nodeCreate.getLocation()), eq(tenantID));
```
Comment (james8hut) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/service/NodeServiceTest.java:274]: Is it possible to add a verify statement saying that 0 calls have been made to nodeRepository save method? (Just to explicitly say in the test that we're definitely not creating a node.)

Comment (yangli2014) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/service/NodeServiceTest.java:274]: sure

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 3]
```diff
@@ -184,6 +183,15 @@ private Node saveNode(NodeCreateDTO request, MonitoringLocation monitoringLocati
 
     @Transactional
     public Node createNode(NodeCreateDTO request, ScanType scanType, String tenantId) {
```
[the remark this answers is not in the document]
Comment (yangli2014) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java:185]: I will look at that after today's demo. It makes sense to me.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 4]
```diff
@@ -183,7 +183,16 @@ private Node saveNode(NodeCreateDTO request, MonitoringLocation monitoringLocati
     }
 
     @Transactional
-    public Node createNode(NodeCreateDTO request, ScanType scanType, String tenantId) {
+    public Node createNode(NodeCreateDTO request, ScanType scanType, String tenantId) throws EntityExistException {
+        if(request.hasManagementIp()) { //Do we really want to create a node without managed IP?
+            Optional<IpInterface> ipInterfaceOpt = ipInterfaceRepository
+                .findByIpAddressAndLocationAndTenantId(InetAddressUtils.getInetAddress(request.getManagementIp()), request.getLocation(), tenantId);
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java:189]: we can remove `ipInterfaceRepository
                .findByIpAddressAndLocationAndTenantId` call from the places where this method is being called. That would save one us one DB query

File: alert/src/main/java/org/opennms/horizon/alertservice/service/routing/EventConsumer.java [hunk 5]
```diff
@@ -49,25 +49,24 @@
...
             EventLog eventLog = EventLog.parseFrom(data);
-
-
             eventLog.getEventsList().forEach(e -> {
                 if (Strings.isNullOrEmpty(e.getTenantId())) {
                     LOG.warn("TenantId is empty, dropping event: {}", e);
                     return;
                 }
 
+                if(e.getNodeId() <= 0) {
```
Comment (james8hut) [inline alert/src/main/java/org/opennms/horizon/alertservice/service/routing/EventConsumer.java:63]: Possibly unrelated to this PR, but can I ask how is it possible to get an event with no node? When I was doing some testing yesterday, I sent an snmpTrap to create a node, the node was created, but the event did not know about the node. I imagine this is a bug?

**Review by james8hut — APPROVED** [review]

File: alert/src/main/java/org/opennms/horizon/alertservice/service/routing/EventConsumer.java (hunk 5 above)
Comment (yangli2014) [reply alert/src/main/java/org/opennms/horizon/alertservice/service/routing/EventConsumer.java:63]: When the first trap event arrives, the isn't a node in the database. So if you test you can create two alerts one with node id is 0

**Pull request merged**